Frameworks/Foundation/NSPointerArray.mm

```diff
@@ -14,146 +14,396 @@
+                _container[lastNonNull] = _container[i];
+            }
+            lastNonNull++;
+        }
+    }
+
+    _count = lastNonNull;
+
+    if (_count < _allocated) {
+        _container = static_cast<void**>(realloc(_container, sizeof(void*) * _count));
+        _allocated = _count;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (instancetype)initWithCoder:(NSCoder*)decoder {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (self = [super init]) {
+        NSUInteger options = 0;
+        options = [decoder decodeIntegerForKey:@"NS.options"];
+        NSUInteger count = [decoder decodeIntegerForKey:@"NS.count"];
+
+        _pointerFunctions = [[NSPointerFunctions alloc] initWithOptions:options];
+
+        for (NSUInteger i = 0; i < count; i++) {
+            NSObject* obj = [decoder decodeObject];
+            [self addPointer:static_cast<void*>(obj)];
+        }
+    }
+    return self;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)encodeWithCoder:(NSCoder*)encoder {
-    UNIMPLEMENTED();
+    if (!self.pointerFunctions.usesStrongWriteBarrier && !self.pointerFunctions.usesWeakReadAndWriteBarriers) {
+        [NSException raise:NSInvalidArgumentException format:@"NSPointerArray has options that cannot be encoded"];
+    }
+
+    NSUInteger options = (self.pointerFunctions.usesStrongWriteBarrier ? NSPointerFunctionsStrongMemory : NSPointerFunctionsWeakMemory);
+    [encoder encodeInteger:options forKey:@"NS.options"];
+    [encoder encodeInteger:_count forKey:@"NS.count"];
+
+    for (NSObject* obj in self) {
+        [encoder encodeObject:obj];
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (id)copyWithZone:(NSZone*)zone {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSPointerArray* copy = [NSPointerArray pointerArrayWithPointerFunctions:self.pointerFunctions];
+    auto acquireFunction = self.pointerFunctions.acquireFunction;
+    copy.count = _count;
+
+    if (acquireFunction) {
+        for (NSUInteger i = 0; i < _count; i++) {
+            if (_container[i]) {
+                copy->_container[i] = acquireFunction(_container[i], sizeFunction, NO);
+            }
+        }
+    } else {
+        memcpy(copy->_container, _container, sizeof(void*) * _count);
+    }
+
+    return copy;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState*)state objects:(id _Nonnull[])stackbuf count:(NSUInteger)len {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (state->state == 0) {
+        state->mutationsPtr = reinterpret_cast<unsigned long*>(self);
+        state->extra[0] = 0;
+        state->state++;
+    }
+
+    if (state->extra[0] > _count) {
+        return 0;
+    }
+
+    NSUInteger slice = MIN(len, _count - state->extra[0]);
+
+    state->itemsPtr = reinterpret_cast<id*>(_container + state->extra[0]);
+    state->extra[0] += slice;
+
+    return slice;
```

Comment:
I added as much comments as possible.
Also, I changed the behaviour of the "isEqual:" and "hash" method in order to match the behaviour of the reference platform.